lib/WebpackOptionsApply.js

```diff
@@ -44,10 +44,36 @@ const DefaultStatsFactoryPlugin = require("./stats/DefaultStatsFactoryPlugin");
 const DefaultStatsPresetPlugin = require("./stats/DefaultStatsPresetPlugin");
 const DefaultStatsPrinterPlugin = require("./stats/DefaultStatsPrinterPlugin");
 
-const { cachedCleverMerge } = require("./util/cleverMerge");
+const { cachedCleverMerge, DELETE } = require("./util/cleverMerge");
 
+/** @typedef {import("../declarations/WebpackOptions").ResolveOptions} ResolveOptions */
 /** @typedef {import("../declarations/WebpackOptions").WebpackOptionsNormalized} WebpackOptions */
 /** @typedef {import("./Compiler")} Compiler */
+/** @typedef {import("./ResolverFactory").EnhancedResolveOptions} EnhancedResolveOptions */
+
+// need to be hoisted on module level for caching identity
+const noByDependency = {
+	byDependency: DELETE,
+	dependencyType: DELETE
+};
+
+/**
+ * @param {EnhancedResolveOptions} enhancedResolveOptions enhanced options
+ * @returns {ResolveOptions} merged options
+ */
+const applyByDependency = enhancedResolveOptions => {
+	if (!enhancedResolveOptions.byDependency) return enhancedResolveOptions;
+	const options = /** @type {ResolveOptions} */ (cachedCleverMerge(
+		enhancedResolveOptions,
+		noByDependency
+	));
+	if (!enhancedResolveOptions.dependencyType) return options;
```

Comment:
```suggestion
	const options = /** @type {ResolveOptions} */ (cachedCleverMerge(
		enhancedResolveOptions,
		noByDependency
	));
	if (!enhancedResolveOptions.byDependency || !enhancedResolveOptions.dependencyType) {
		return options;
	}
```

We need to delete the properties in any case.